testutils: make MoveRangeLeaseNonCooperatively() work with leader leases

iskettaneh opened the pull request:
In leader leases, trying to non cooperatively take the lease might return `leader lease is not held locally, cannot determine validity`. This commit allows (TestCluster).MoveRangeLeaseNonCooperatively() to work with leader leases by forcing another leader to step up when we receive an error when taking the lease non cooperatively.

References: #136744

Release Note: None

iskettaneh committed:
testcluster: make MoveRangeLeaseNonCooperatively() work with leader leases

In leader leases, trying to non cooperatively take the lease might return
`leader lease is not held locally, cannot determine validity`. This commit
allows (TestCluster).MoveRangeLeaseNonCooperatively() to work with leader
leases by forcing another leader to step up when we receive an error when
taking the lease non cooperatively.

References: #136744

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/137053)
<!-- Reviewable:end -->

arulajmani reviewed (commented):
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137053)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh)

___
*[`pkg/testutils/testcluster/testcluster.go` line 1245 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODlF3Zv6DOShdXJlBVM:-ODlF3Zv6DOShdXJlBVN:b-k1gtjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccce3261d54b4b6a068ef1784d95c428525fd52/pkg/testutils/testcluster/testcluster.go#L1245)):*
> ```Go
> 			if prevLease.Type() == roachpb.LeaseLeader {
> 				// With leader leases, the replica might not be able to determine the
> 				// validity of the lease. In this case, we want for another leader to
> ```

Do we care about the validity of the lease though? Regardless of whether the lease is considered valid or not, it should point to the leader, right?

Could this `prevLease.Type() == roachpb.LeaseLeader` logic be below, in the "Is the lease in the right place?" part of the code?
___
*[`pkg/testutils/testcluster/testcluster.go` line 1296 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODlEXd59I9H5Ywk6if9:-ODlEXd59I9H5Ywk6ifA:bvl1b2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccce3261d54b4b6a068ef1784d95c428525fd52/pkg/testutils/testcluster/testcluster.go#L1296)):*
> ```Go
> 
> 			if curR.RaftStatus().RaftState == raftpb.StateLeader &&
> 				curR.RaftStatus().LeadSupportUntil.GoTime().After(manual.Now()) {
> ```

Why do we care whether the leader is fortified or not?
___
*[`pkg/testutils/testcluster/testcluster.go` line 1297 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODlEUA206nQS8En8oLX:-ODlEUA206nQS8En8oLY:bchu8fa) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccce3261d54b4b6a068ef1784d95c428525fd52/pkg/testutils/testcluster/testcluster.go#L1297)):*
> ```Go
> 			if curR.RaftStatus().RaftState == raftpb.StateLeader &&
> 				curR.RaftStatus().LeadSupportUntil.GoTime().After(manual.Now()) {
> 				log.Infof(ctx, "Current fortified leader is %v at term: %d",
> ```

nit: log lines shouldn't start with capital letters.
___
*[`pkg/testutils/testcluster/testcluster.go` line 1309 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODlEIpRE551JCK3VkPW:-ODlEIpRE551JCK3VkPX:bivbcai) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccce3261d54b4b6a068ef1784d95c428525fd52/pkg/testutils/testcluster/testcluster.go#L1309)):*
> ```Go
> 			break
> 		}
> 		time.Sleep(100 * time.Millisecond)
> ```

Typically, we don't want to have raw sleeps. Instead, we should use `SucceedsSoon` or a similar function.
___
*[`pkg/testutils/testcluster/testcluster.go` line 1312 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODlEeif6k_BBlGeqA9i:-ODlEeif6k_BBlGeqA9j:b-om9lhe) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccce3261d54b4b6a068ef1784d95c428525fd52/pkg/testutils/testcluster/testcluster.go#L1312)):*
> ```Go
> 	}
> 
> 	// Block store liveness messages to the current leader.
> ```

Could you remind me why we need to block store liveness messages to the leader and simply incrementing the manual clock doesn't work?
___
*[`pkg/testutils/testcluster/testcluster.go` line 1335 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODlEr-i5D-3D7FzRggr:-ODlEr-i5D-3D7FzRggs:b-gjwpv0) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccce3261d54b4b6a068ef1784d95c428525fd52/pkg/testutils/testcluster/testcluster.go#L1335)):*
> ```Go
> 		}
> 		log.Infof(ctx, "waiting for leader step down")
> 		time.Sleep(100 * time.Millisecond)
> ```

Like above, we'd typically want this to be in a succeeds soon instead of a raw for/sleep/break.


<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137053)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/testutils/testcluster/testcluster.go` line 1245 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODlF3Zv6DOShdXJlBVM:-ODltD3F5kgaGjWELlPM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccce3261d54b4b6a068ef1784d95c428525fd52/pkg/testutils/testcluster/testcluster.go#L1245)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Do we care about the validity of the lease though? Regardless of whether the lease is considered valid or not, it should point to the leader, right?

Could this `prevLease.Type() == roachpb.LeaseLeader` logic be below, in the "Is the lease in the right place?" part of the code?
</blockquote></details>

Done.
___
*[`pkg/testutils/testcluster/testcluster.go` line 1296 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODlEXd59I9H5Ywk6if9:-ODlsNL60Tah9Oe--ueV:bqk4ei0) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccce3261d54b4b6a068ef1784d95c428525fd52/pkg/testutils/testcluster/testcluster.go#L1296)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Why do we care whether the leader is fortified or not?
</blockquote></details>

I guess the wording could change a bit. But basically, this function relies on the fact the leader would step down if it's not supported by a majority. However, that code path is only exercised if store liveness is enabled.

Do you think I should reword the function name to `StepDownLeaderWithStoreLivenessEnabled()`?
___
*[`pkg/testutils/testcluster/testcluster.go` line 1309 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODlEIpRE551JCK3VkPW:-ODlve3A4Li7RlYPDTgX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccce3261d54b4b6a068ef1784d95c428525fd52/pkg/testutils/testcluster/testcluster.go#L1309)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Typically, we don't want to have raw sleeps. Instead, we should use `SucceedsSoon` or a similar function.
</blockquote></details>

Done.
___
*[`pkg/testutils/testcluster/testcluster.go` line 1312 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODlEeif6k_BBlGeqA9i:-ODlrY4T6tskuY6worxb:bnemrvx) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccce3261d54b4b6a068ef1784d95c428525fd52/pkg/testutils/testcluster/testcluster.go#L1312)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Could you remind me why we need to block store liveness messages to the leader and simply incrementing the manual clock doesn't work?
</blockquote></details>

I want the leader to step down. I thought the way to do it is to (1) prevent the leader from getting more support from other nodes, and (2) advance the clock. This way, the leader would step down on the next CheckQuorum
___
*[`pkg/testutils/testcluster/testcluster.go` line 1335 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODlEr-i5D-3D7FzRggr:-ODlvfI9Aik2C8hbl_bt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccce3261d54b4b6a068ef1784d95c428525fd52/pkg/testutils/testcluster/testcluster.go#L1335)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Like above, we'd typically want this to be in a succeeds soon instead of a raw for/sleep/break.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137053)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh)

___
*[`pkg/testutils/testcluster/testcluster.go` line 1296 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODlEXd59I9H5Ywk6if9:-ODs0_k47N_4C4wfe4k3:b-vy6dsu) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccce3261d54b4b6a068ef1784d95c428525fd52/pkg/testutils/testcluster/testcluster.go#L1296)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

I guess the wording could change a bit. But basically, this function relies on the fact the leader would step down if it's not supported by a majority. However, that code path is only exercised if store liveness is enabled.

Do you think I should reword the function name to `StepDownLeaderWithStoreLivenessEnabled()`?
</blockquote></details>

Instead of needing to figure out who the leader is, can we instead pass that information from the caller who knows this by virtue of having a leader lease? 
___
*[`pkg/testutils/testcluster/testcluster.go` line 1312 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODlEeif6k_BBlGeqA9i:-ODrzQ9N4V-sbqb14QxI:b-rlih9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccce3261d54b4b6a068ef1784d95c428525fd52/pkg/testutils/testcluster/testcluster.go#L1312)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

I want the leader to step down. I thought the way to do it is to (1) prevent the leader from getting more support from other nodes, and (2) advance the clock. This way, the leader would step down on the next CheckQuorum
</blockquote></details>

I see, so this guards us against the race where `StorLiveness` is able to heartbeat at the advanced timestamp. Makes sense, thanks for explaining. 
___
*[`pkg/testutils/testcluster/testcluster.go` line 1254 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODs0RLgAX11N9nIPdxO:-ODs0RLgAX11N9nIPdxP:b-ibeqip) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcb49fd4ea9d921a1c33183f2744d8226653c6e2/pkg/testutils/testcluster/testcluster.go#L1254)):*
> ```Go
> 		// Is the lease in the right place?
> 		if newLease.Replica.StoreID != dest.StoreID {
> 			if prevLease.Type() == roachpb.LeaseLeader {
> ```

Shouldn't this be `newLease` instead?
___
*[`pkg/testutils/testcluster/testcluster.go` line 1256 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODrz25L0Z0DsC1q6mvW:-ODrz25L0Z0DsC1q6mvX:b-befo7b) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcb49fd4ea9d921a1c33183f2744d8226653c6e2/pkg/testutils/testcluster/testcluster.go#L1256)):*
> ```Go
> 			if prevLease.Type() == roachpb.LeaseLeader {
> 				// With leader leases, the replica might not be able to determine the
> 				// validity of the lease. In this case, we want for another leader to
> ```

This comment is stale. In particular, the validity bit doesn't make much sense. Let's reword it.
___
*[`pkg/testutils/testcluster/testcluster.go` line 1274 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODs0kzdDNhTkCyDIk3W:-ODs0kzdDNhTkCyDIk3X:b7cz7ro) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcb49fd4ea9d921a1c33183f2744d8226653c6e2/pkg/testutils/testcluster/testcluster.go#L1274)):*
> ```Go
> // StepDownLeaderWithStoreLivenessEnabled withdraws store liveness support from
> // the leader, and waits for it to step down.
> func (tc *TestCluster) StepDownLeaderWithStoreLivenessEnabled(
> ```

nit: consider renaming this to `ensureLeaderStepDown`. The `StoreLivenessEnabled` bit can be tucked away as a comment on the function instead.
___
*[`pkg/testutils/testcluster/testcluster.go` line 1281 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODs-fP94Q0wK1FroDRR:-ODs-fP94Q0wK1FroDRS:b30nkq6) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcb49fd4ea9d921a1c33183f2744d8226653c6e2/pkg/testutils/testcluster/testcluster.go#L1281)):*
> ```Go
> 	var leaderReplica *kvserver.Replica
> 
> 	// Wait until a leader gets elected.
> ```

Similar to my comment below, let's use `testutils.SucceedsSoon` here instead.
___
*[`pkg/testutils/testcluster/testcluster.go` line 1331 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODs-26l728hecOKIPmZ:-ODs-26l728hecOKIPm_:b9bupq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcb49fd4ea9d921a1c33183f2744d8226653c6e2/pkg/testutils/testcluster/testcluster.go#L1331)):*
> ```Go
> 
> 	// Wait for the leader to step down.
> 	r = retry.StartWithCtx(ctx, retry.Options{
> ```

Why this instead of `testutils.SucceedsSoon`? Is it because you don't have a `testing.T` handy? If so, there's not that many callers to `MoveRangeLeaseNonCooperatively` -- maybe we can clean this up in a previous commit (and while we're there, replace the `retry.ForDuration` as well). 


<!-- Sent from Reviewable.io -->

iskettaneh committed:
testcluster: pass testing.T to MoveRangeLeaseNonCooperatively()

This commit passes testing.T to MoveRangeLeaseNonCooperatively() which
allows it to use testutils like SucceedsWithin.

References: #137235

Release note: None

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137053)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/testutils/testcluster/testcluster.go` line 1296 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODlEXd59I9H5Ywk6if9:-ODvMmJ_CQc9oGWerYFS:b107413) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccce3261d54b4b6a068ef1784d95c428525fd52/pkg/testutils/testcluster/testcluster.go#L1296)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Instead of needing to figure out who the leader is, can we instead pass that information from the caller who knows this by virtue of having a leader lease?
</blockquote></details>

This didn't work since there could be a leader change between the time we send who the leader is to step down.
___
*[`pkg/testutils/testcluster/testcluster.go` line 1312 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODlEeif6k_BBlGeqA9i:-ODvERWF5KtSLDR0C-z1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccce3261d54b4b6a068ef1784d95c428525fd52/pkg/testutils/testcluster/testcluster.go#L1312)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I see, so this guards us against the race where `StorLiveness` is able to heartbeat at the advanced timestamp. Makes sense, thanks for explaining.
</blockquote></details>

Done.
___
*[`pkg/testutils/testcluster/testcluster.go` line 1254 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODs0RLgAX11N9nIPdxO:-ODvMXi92DyeLU0aDwTD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcb49fd4ea9d921a1c33183f2744d8226653c6e2/pkg/testutils/testcluster/testcluster.go#L1254)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Shouldn't this be `newLease` instead?
</blockquote></details>

Done.
___
*[`pkg/testutils/testcluster/testcluster.go` line 1256 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODrz25L0Z0DsC1q6mvW:-ODvMXRL5e4YzAVJY2Kd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcb49fd4ea9d921a1c33183f2744d8226653c6e2/pkg/testutils/testcluster/testcluster.go#L1256)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

This comment is stale. In particular, the validity bit doesn't make much sense. Let's reword it.
</blockquote></details>

Done.
___
*[`pkg/testutils/testcluster/testcluster.go` line 1281 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODs-fP94Q0wK1FroDRR:-OE-mhqB3yk9gxtTT47R:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcb49fd4ea9d921a1c33183f2744d8226653c6e2/pkg/testutils/testcluster/testcluster.go#L1281)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Similar to my comment below, let's use `testutils.SucceedsSoon` here instead.
</blockquote></details>

Done.
___
*[`pkg/testutils/testcluster/testcluster.go` line 1331 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-ODs-26l728hecOKIPmZ:-OE-mieoDw0W5B5uIPCj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcb49fd4ea9d921a1c33183f2744d8226653c6e2/pkg/testutils/testcluster/testcluster.go#L1331)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Why this instead of `testutils.SucceedsSoon`? Is it because you don't have a `testing.T` handy? If so, there's not that many callers to `MoveRangeLeaseNonCooperatively` -- maybe we can clean this up in a previous commit (and while we're there, replace the `retry.ForDuration` as well).
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 7 of 7 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-:-OEYjmYo5mb5SFW_p4Wx:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @iskettaneh)

___
*[`pkg/testutils/testcluster/testcluster.go` line 1289 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-OEYjAP25ZQX1Iv7o6mO:-OEYjAP25ZQX1Iv7o6mP:b-ni9i6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/711e68dc10bed84e70ea048f3a763f57279fdd2d/pkg/testutils/testcluster/testcluster.go#L1289)):*
> ```Go
> 		// Is the lease in the right place?
> 		if newLease.Replica.StoreID != dest.StoreID {
> 			if newLease.Type() == roachpb.LeaseLeader {
> ```

I still think we should be deliberate about establishing a leader in the right place instead first when we want a leader lease instead of acquiring a lease and then checking if it's in the right place. I know this is just a testing patch, so maybe we don't block this PR over this, but let's at least add a TODO for that approach before merging.
___
*[`pkg/testutils/testcluster/testcluster.go` line 1292 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-OEYhg3T9yYxGRTior1D:-OEYhg3T9yYxGRTior1E:b-vimoso) ([raw file](https://github.com/cockroachdb/cockroach/blob/711e68dc10bed84e70ea048f3a763f57279fdd2d/pkg/testutils/testcluster/testcluster.go#L1292)):*
> ```Go
> 				// With leader leases, we want to current leader to step down to give
> 				// the new leader a chance to acquire the lease.
> 				if leaderLeaseErr :=
> ```

nit: s/leaderLeaseErr/err/g
___
*[`pkg/testutils/testcluster/testcluster.go` line 1318 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-OEYjX7L31HvISLBPyzI:-OEYjX7L31HvISLBPyzJ:b-gtu8h3) ([raw file](https://github.com/cockroachdb/cockroach/blob/711e68dc10bed84e70ea048f3a763f57279fdd2d/pkg/testutils/testcluster/testcluster.go#L1318)):*
> ```Go
> 	var leaderReplica *kvserver.Replica
> 
> 	// Wait until a leader gets elected.
> ```

Consider expanding this comment to talk about why we need to wait for a leader to get elected. Separately, we're also storing who the leader is here, which deserves to be called out.
___
*[`pkg/testutils/testcluster/testcluster.go` line 1341 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-OEYgdfhErgph-i5CwHy:-OEYgdfhErgph-i5CwHz:bafl20b) ([raw file](https://github.com/cockroachdb/cockroach/blob/711e68dc10bed84e70ea048f3a763f57279fdd2d/pkg/testutils/testcluster/testcluster.go#L1341)):*
> ```Go
> 		}
> 		// At this point we have iterated over all nodes in the cluster, if we
> 		// haven't found the fortified leader, we should wait for a bit and retry.
> ```

nit: this comment needs an update now that we're not checking whether the leader is fortified or not. Should it simply say "if we haven't found a leader, wait for a bit for one to step up"


<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137053)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/testutils/testcluster/testcluster.go` line 1289 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-OEYjAP25ZQX1Iv7o6mO:-OFc8g6H1-j7ME5i_8vh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/711e68dc10bed84e70ea048f3a763f57279fdd2d/pkg/testutils/testcluster/testcluster.go#L1289)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I still think we should be deliberate about establishing a leader in the right place instead first when we want a leader lease instead of acquiring a lease and then checking if it's in the right place. I know this is just a testing patch, so maybe we don't block this PR over this, but let's at least add a TODO for that approach before merging.
</blockquote></details>

Done.
___
*[`pkg/testutils/testcluster/testcluster.go` line 1318 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/137053#-OEYjX7L31HvISLBPyzI:-OFc82wA4JecxazOoNmR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/711e68dc10bed84e70ea048f3a763f57279fdd2d/pkg/testutils/testcluster/testcluster.go#L1318)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Consider expanding this comment to talk about why we need to wait for a leader to get elected. Separately, we're also storing who the leader is here, which deserves to be called out.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

iskettaneh commented:
TFTR!

bors r+
